services/shell/copy_rpc.go

```diff
@@ -0,0 +1,478 @@
+package shell
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"path/filepath"
+	"sync"
+	"time"
+
+	pb "go.viam.com/api/service/shell/v1"
+	"go.viam.com/utils"
+	"go.viam.com/utils/rpc"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// copy_rpc supports RPC based copy operations and is agnostic to where files
+// originated from beyond the RPC boundary. The main testing for this is done
+// in the CLI package since it's just easier to coordinate over there.
+
+// newCopyFileFromMachineFactory returns a factory that can make FileCopiers for
+// the Writer/Server side of a CopyFrom.
+func newCopyFileFromMachineFactory(
+	srv pb.ShellService_CopyFilesFromMachineServer, preserve bool,
+) FileCopyFactory {
+	return &copyFileFromMachineFactory{srv: srv, preserve: preserve}
+}
+
+type copyFileFromMachineFactory struct {
+	once     bool
```

Comment:
[q] In what case would you make multiple file copiers/why the need for a guard here?

Reply:
There shouldn't be a case, so this is here as an invariant. I can't prevent the caller from calling this twice and since it's reusable, it feels like it makes sense to prevent an invalid initial `CopyFilesFromMachineResponse` from being sent. Happy to remove if it feels too defensive/paranoid though

Reply:
Having the defensive case seems fine to me; I just couldn't tell if I was missing an obvious case where a CLI user might trigger a double call. Thanks for the context.